gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java

```diff
@@ -0,0 +1,90 @@
+    private RecommendedProductDependencyMerger() { }
+
+    public static RecommendedProductDependency mergeRecommendedProductDependencies(
+            RecommendedProductDependency dep1, RecommendedProductDependency dep2) {
+        // Ensure they are valid
+        Arrays.asList(dep1, dep2).forEach(RecommendedProductDependency::isValid);
+        if (!dep1.getProductGroup().equals(dep2.getProductGroup())) {
+            throw new IllegalArgumentException(String.format("Product groups differ: '%s' and '%s'",
+                dep1.getProductGroup(), dep2.getProductGroup()));
+        }
+        if (!dep1.getProductName().equals(dep2.getProductName())) {
+            throw new IllegalArgumentException(String.format("Product names differ: '%s' and '%s'",
+                dep1.getProductName(), dep2.getProductName()));
+        }
+
+        OrderableSlsVersion minimumVersion = Stream
+                .of(
+                        OrderableSlsVersion.valueOf(dep1.getMinimumVersion()),
+                        OrderableSlsVersion.valueOf(dep2.getMinimumVersion()))
+                .max(VersionComparator.INSTANCE)
+                .orElseThrow(() -> new RuntimeException("Impossible"));
+
+        Optional<MaximumVersion> maximumVersion = Stream
+                .of(dep1.getMaximumVersion(), dep2.getMaximumVersion())
+                .filter(Objects::nonNull)
+                .map(MaximumVersion::valueOf)
+                .min(MaximumVersionComparator.INSTANCE);
+
+        // Sanity check: min has to be <= max
+        if (!satisfiesMaxVersion(maximumVersion, minimumVersion)) {
+            throw new IllegalArgumentException(String.format(
+            "Could not merge recommended product dependencies as their version ranges do not overlap: '%s' and '%s'. "
+                        + "Merged min: %s, merged max: %s",
+                dep1, dep2, minimumVersion, maximumVersion));
+        }
+
+        // Recommended version. Check that it matches the inferred min and max.
+        // If none of them do, then pick the min version.
+        Optional<OrderableSlsVersion> recommendedVersion = Stream
+                .of(dep1.getRecommendedVersion(), dep2.getRecommendedVersion())
+                .filter(Objects::nonNull)
+                .map(OrderableSlsVersion::valueOf)
+                .filter(version -> VersionComparator.INSTANCE.compare(version, minimumVersion) >= 0
+                        && satisfiesMaxVersion(maximumVersion, version))
+                .max(VersionComparator.INSTANCE);
+
+        RecommendedProductDependency result = new RecommendedProductDependency();
+        result.setMinimumVersion(minimumVersion.toString());
+        maximumVersion.map(Objects::toString).ifPresent(result::setMaximumVersion);
+        recommendedVersion.map(Objects::toString).ifPresent(result::setRecommendedVersion);
+        result.setProductGroup(dep1.getProductGroup());
+        result.setProductName(dep1.getProductName());
+        // Verify validity
```

Comment:
nit: this comment is redundant since the next line literally says `result.isValid();` . if anything, maybe consider renaming `isValid` to something like `verifyIsValid`. (does it throw?)